src/manifest_store.rs

```diff
@@ -161,8 +180,77 @@ impl StoredManifest {
         Ok(&self.manifest.core)
     }
 
+    /// Create a new checkpoint from the current state. This only creates the checkpoint
+    /// struct, but does not persist it in the manifest. The latter can be done with
+    /// [`Self::write_checkpoint`].
+    pub(crate) fn new_checkpoint(
+        &self,
+        checkpoint_id: Uuid,
+        options: &CheckpointOptions,
+    ) -> Result<Checkpoint, SlateDBError> {
+        let expire_time = options.lifetime.map(|l| SystemTime::now() + l);
+        let db_state = self.db_state();
+        let manifest_id = match options.source {
+            Some(source_checkpoint_id) => {
+                let Some(source_checkpoint) = db_state.find_checkpoint(&source_checkpoint_id)
+                else {
+                    return Err(InvalidDBState);
+                };
+                source_checkpoint.manifest_id
+            }
+            None => {
+                if !db_state.initialized {
+                    return Err(InvalidDBState);
+                }
+                self.id()
+            }
+        };
+        Ok(Checkpoint {
+            id: checkpoint_id,
+            manifest_id,
+            expire_time,
+            create_time: SystemTime::now(),
+        })
+    }
+
+    /// Write the given checkpoint to the manifest,
+    pub(crate) async fn write_checkpoint(
+        &mut self,
+        checkpoint: Checkpoint,
+    ) -> Result<(), SlateDBError> {
+        self.maybe_apply_db_state_update(|stored_manifest| {
+            // TODO: Do we need to verify that the manifest is still available
+            //       before writing?
+            let mut updated_db_state = stored_manifest.db_state().clone();
+            updated_db_state.checkpoints.push(checkpoint.clone());
+            Ok(Some(updated_db_state))
+        })
+        .await
+    }
+
+    pub(crate) async fn write_new_checkpoint(
```

Comment:
Feels kinda funny to have write checkpoint and write new checkpoint as two different methods. I wasn't sure what each one did. At a minimum, could you add some docs to this method to explain how it differs from the other? IIUC, it's meant to be used when you want to checkpoint from the current manifest, while write checkpoint is meant to be used when a checkpoint has already been created (potentially from an older manifest).

Reply:
The original RFC was a little awkward because we had to create the uninitialized checkpoint first, write the reference to it in the clone manifest, then add it to the parent. That is why I initially had a separate path for checkpoint creation. The logic was trying to handle the case when the user wanted to use the latest state and did not have a handy checkpoint ID already. I modified the protocol to instead create an ephemeral checkpoint against the latest state. Then we always have a source checkpoint when writing the final clone checkpoint in the parent manifest which means we don't need two paths.